File: Exercism/haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import qualified Data.Char as Char

import Data.IntMap.Lazy (IntMap)
import qualified Data.IntMap.Lazy as IntMap


type Counter =  IntMap Int


incrementCounter :: IntMap.Key -> Counter -> Counter
incrementCounter = IntMap.alter setCounter
  where
    setCounter (Just x) = Just (x+1)
    setCounter Nothing = Just 1

anagramsFor :: String -> [String] -> [String]
anagramsFor word = filter isAnagram
  where
    isAnagram = (&&) <$> (toLower word /=) . toLower <*> (anagramSignature word ==) . anagramSignature
    toLower = map Char.toLower
    anagramSignature = foldr (incrementCounter . Char.ord . Char.toLower) IntMap.empty

File: Exercism/python/anagram/anagram.py
from collections import Counter


def find_anagrams(word, candidates):
    _word = word.lower()
    _letters = Counter(_word)

    def is_anagram(candidate):
        return candidate != _word and Counter(candidate) == _letters

    return [c for c in candidates if is_anagram(c.lower())]
